### DISCUSSION TITLE
What does `committed_state` mean?

### DISCUSSION THREAD

**Discussion by ehdgua01** [sqlalchemy/sqlalchemy#7344] [Usage Questions] [answered]
I can't find any information about the `committed_state` attribute of `InstanceState`

**Comment by zzzeek** [comment, marked as answer]
it's a dictionary that refers to the values for attributes that were last loaded and/or persisted from the database and were then possibly replaced with a new value in that instance's `__dict__`.

when you load an object from the DB, committed_state is empty.  `obj.__dict__` is populated with values that were loaded from the row.   When you alter an attribute , the value that's already there, which is considered to be what was loaded from the database, is added to the committed_state dictionary.  then the incoming value being set in Python code is placed in ``obj.__dict__``.

When the unit of work then seeks to flush changes on the object, it looks at the "history" of each attribute to see if each one changed.   this is calculated by looking in ``committed_state`` for a value, then comparing that value to what's in `obj.__dict__`.   if these don't match, then the attribute has changed, and in the unit of work you will see that column in an UPDATE statement.

The effects of committed_state therefore are easier to reason about if you use [the history API](https://docs.sqlalchemy.org/en/14/orm/internals.html#sqlalchemy.orm.AttributeState) which will interpret committed_state accordingly.

put another way, committed_state should not have been public.

**Reply by donghyeon-kim98** [reply]
thank you @zzzeek 
Can the attribute disappear? I tried to use the attribute to solve my problem.

**Reply by zzzeek** [reply]
you shouldn't write to this attribute or alter it.   what I woudl do in 2.0 would be to make it read-only if i get around to it.
